File: sdk/python/awa/runtime/workflow_engine.py
"""
AWA Workflow Engine
Executes AWA workflow definitions
"""

import os
from datetime import datetime
from enum import Enum
from typing import Any
from uuid import uuid4

def format_iso_duration(seconds: float) -> str:
    """Format seconds as ISO 8601 duration string (PTnHnMnS)"""
    if seconds < 0:
        seconds = 0
    
    days = int(seconds // 86400)
    hours = int((seconds % 86400) // 3600)
    minutes = int((seconds % 3600) // 60)
    secs = seconds % 60
    
    result = "P"
    if days > 0:
        result += f"{days}D"
    
    result += "T"
    if hours > 0:
        result += f"{hours}H"
    if minutes > 0:
        result += f"{minutes}M"
    if secs > 0 or result == "PT":
        result += f"{secs:.1f}S"
    
    return result

from awa.types import Workflow, Activity, ActorType, DecisionNode, WasteCategory
from awa.runtime.token import Token, TokenStatus
from awa.runtime.context_manager import ContextManager
from awa.runtime.actors import SoftwareAgent, AIAgent, HumanAgent, RobotAgent
from awa.runtime.decision_evaluator import DecisionEvaluator
from awa.validator import validate_workflow_integrity


class EngineStatus(str, Enum):
    """Workflow engine status"""
    IDLE = "idle"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    WAITING_HUMAN = "waiting_human"


class WorkflowEngine:
    """
    Executes AWA workflow definitions.
    """
    
    def __init__(
        self,
        workflow: Workflow,
        gemini_api_key: str | None = None,
        verbose: bool = False
    ):
        self.workflow = workflow
        self.gemini_api_key = gemini_api_key or os.environ.get("GEMINI_API_KEY")
        self.verbose = verbose
        
        self._context_manager = ContextManager(workflow.contexts)
        self._tokens: dict[str, Token] = {}
        self._status = EngineStatus.IDLE
        self._activities_by_id: dict[str, Activity] = {
            str(a.id): a for a in workflow.activities
        }
        self._decision_nodes_by_id: dict[str, DecisionNode] = {
            str(d.id): d for d in workflow.decision_nodes
        }
        
        # Initialize actors
        self._actors = {
            ActorType.application: SoftwareAgent(),
            ActorType.ai_agent: AIAgent(self.gemini_api_key),
            ActorType.human: HumanAgent(),
            ActorType.robot: RobotAgent(),
        }
        self._decision_evaluator = DecisionEvaluator()
        
        # Ensure we use strings for comparison in node ID search later
        # Find start activity initially
        self._start_activity = self._find_start_activity()
        
        # Validate workflow
        if not validate_workflow_integrity(workflow):
            raise ValueError("Invalid workflow definition")
    
    @property
    def status(self) -> EngineStatus:
        """Get engine status"""
        return self._status
    
    @property
    def tokens(self) -> list[Token]:
        """Get all tokens"""
        return list(self._tokens.values())
    
    def run(self, initial_data: dict[str, Any] | None = None) -> dict[str, Any]:
        """
        Execute the workflow from start to end.
        """
        self._status = EngineStatus.RUNNING
        
        # Find start activity
        if not self._start_activity:
            self._status = EngineStatus.FAILED
            raise ValueError("No start activity found")
        
        # Create initial token
        token = Token(
            activity_id=str(self._start_activity.id),
            initial_data=initial_data,
            workflow_id=str(self.workflow.id)
        )
        self._tokens[token.id] = token
        
        if self.verbose:
            print(f"[Engine] Starting workflow: {self.workflow.name}")
            print(f"[Engine] Token created: {token.id}")
        
        try:
            # Execute until completion
            while self._has_active_tokens():
                self._execute_step()
            
            self._status = EngineStatus.COMPLETED
            
            return {
                "status": "completed",
                "tokens": [t.to_dict() for t in self._tokens.values()],
                "contexts": self._context_manager.get_all_data()
            }
            
        except Exception as e:
            self._status = EngineStatus.FAILED
            raise
    
    def _find_start_activity(self) -> Activity | None:
        """Find the start activity (one with no incoming edges)"""
        target_ids = {e.target_id for e in self.workflow.edges}
        
        for activity in self.workflow.activities:
            if activity.id not in target_ids:
                return activity
        
        return self.workflow.activities[0] if self.workflow.activities else None
    
    def _has_active_tokens(self) -> bool:
        """Check if there are active tokens"""
        return any(
            t.status == TokenStatus.ACTIVE 
            for t in self._tokens.values()
        )
    
    def _execute_step(self) -> None:
        """Execute one step for all active tokens"""
        active_tokens = [
            t for t in self._tokens.values() 
            if t.status == TokenStatus.ACTIVE
        ]
        
        for token in active_tokens:
            node_id = str(token.activity_id)
            start_time = datetime.now()
            
            # Check if it's an activity
            activity = self._activities_by_id.get(node_id)
            if activity:
                if self.verbose:
                    print(f"[Engine] Executing Activity: {activity.name}")
                
                try:
                    result = self._execute_activity(activity, token)
                    token.merge_data(result)
                    
                    end_time = datetime.now()
                    duration_secs = (end_time - start_time).total_seconds()
                    process_time = format_iso_duration(duration_secs)
                    
                    analytics = {
                        "process_time": process_time,
                        "cycle_time": process_time,
                        "lead_time": process_time,
                        "value_added": getattr(activity.analytics, "value_added", True) if activity.analytics else True,
                        "waste_categories": []
                    }

                    # Check for waiting (Human tasks)
                    if result.get("_requires_human_action"):
                        analytics["waste_categories"].append(WasteCategory.waiting.value)
                        token.update_status(TokenStatus.WAITING, analytics=analytics)
                        token.set_data("_waiting_since", end_time.isoformat())
                        continue

                    next_node = self._find_next_node(str(activity.id))
                    if next_node:
                        token.move(str(next_node), analytics=analytics)
                    else:
                        token.update_status(TokenStatus.COMPLETED, analytics=analytics)
                
                except Exception as e:
                    end_time = datetime.now()
                    duration_secs = (end_time - start_time).total_seconds()
                    analytics = {
                        "process_time": format_iso_duration(duration_secs),
                        "waste_categories": [WasteCategory.defects.value],
                        "error_rate": 1.0
                    }
                    token.update_status(TokenStatus.FAILED, analytics=analytics)
                    if self.verbose:
                        print(f"[Engine] Activity Failed: {activity.name} - {str(e)}")
                continue
                
            # Check if it's a decision node
            decision = self._decision_nodes_by_id.get(node_id)
            if decision:
                if self.verbose:
                    print(f"[Engine] Evaluating Decision: {decision.name}")
                
                edge_id = self._decision_evaluator.evaluate(
                    decision, 
                    token.context_data
                )
                
                if edge_id:
                    next_node = self._find_target_by_edge(str(edge_id))
                    if next_node:
                        token.move(str(next_node))
                    else:
                        token.update_status(TokenStatus.FAILED)
                else:
                    token.update_status(TokenStatus.FAILED)
                continue
            
            # Unknown node
            token.update_status(TokenStatus.FAILED)
    
    def _execute_activity(
        self,
        activity: Activity,
        token: Token
    ) -> dict[str, Any]:
        """Execute an activity based on actor type"""
        actor = self._actors.get(activity.actor_type)
        if actor:
            if self.verbose:
                print(f"  [{activity.actor_type.name}] {activity.name}")
            return actor.execute(activity, token)
        
        return {"_completed": True}
    
    def _find_next_node(self, source_id: str) -> str | None:
        """Find next node ID via edges"""
        for edge in self.workflow.edges:
            if str(edge.source_id) == str(source_id):
                return str(edge.target_id)
        return None

    def _find_target_by_edge(self, edge_id: str) -> str | None:
        """Find target node ID for a given edge ID"""
        for edge in self.workflow.edges:
            if str(edge.id) == str(edge_id):
                return str(edge.target_id)
        return None

    def resume_token(self, token_id: str, output: dict[str, Any]) -> bool:
        """Resume a token that was waiting for human input"""
        token = self._tokens.get(token_id)
        if not token or token.status != TokenStatus.WAITING:
            return False

        waiting_since = token.get_data("_waiting_since")
        analytics = None

        if waiting_since:
            wait_start = datetime.fromisoformat(waiting_since)
            wait_duration = (datetime.now() - wait_start).total_seconds()
            analytics = {
                "wait_time": format_iso_duration(wait_duration),
                "waste_categories": [WasteCategory.waiting.value]
            }

        token.merge_data(output)
        token.update_status(TokenStatus.ACTIVE, analytics=analytics)

        if self._status == EngineStatus.WAITING_HUMAN:
            self._status = EngineStatus.RUNNING

        return True

File: sdk/typescript/src/runtime/workflow_engine.ts
/**
 * AWA Workflow Engine
 * Executes AWA workflow definitions with full actor and decision support
 */

import { UUID, Workflow, Activity, Edge, Role, DecisionNode } from '../types';
import { Token } from './token';
import { ContextManager } from './context_manager';
import { validate_workflow_integrity } from '../validator';
import { Actor } from './actors/actor';
import { SoftwareAgent } from './actors/software_agent';
import { AIAgent } from './actors/ai_agent';
import { RobotAgent, RobotConfig } from './actors/robot_agent';
import { HumanAgent } from './actors/human_agent';
import { DecisionEvaluator, evaluate_decision } from './decision_evaluator';
import { calculate_duration } from './duration_utils';
import { WasteCategory, Analytics } from '../types';

export type EngineStatus = 'running' | 'paused' | 'completed' | 'failed' | 'idle' | 'waiting_human';

export interface WorkflowEngineOptions {
    /** Gemini API key for AI agents */
    gemini_api_key?: string;
    /** @deprecated Use gemini_api_key instead */
    geminiApiKey?: string;
    /** Role definitions */
    roles?: Role[];
    /** Robot configuration */
    robot_config?: RobotConfig;
    /** Wait for human tasks to complete before proceeding */
    wait_for_human_tasks?: boolean;
    /** Enable verbose logging */
    verbose?: boolean;
}

export class WorkflowEngine {
    private workflow: Workflow;
    private tokens: Token[];
    private contextManager: ContextManager;
    private status: EngineStatus;
    private activityMap: Map<UUID, Activity>;
    private edgeMap: Map<UUID, Edge[]>; // Source ID -> Edges
    private decisionNodeMap: Map<UUID, DecisionNode>;
    private options: WorkflowEngineOptions;
    private roleMap: Map<UUID, Role>;
    private decisionEvaluator: DecisionEvaluator;

    constructor(workflowDef: Workflow, options: WorkflowEngineOptions = {}) {
        // Validate workflow integrity
        const validation = validate_workflow_integrity(workflowDef);
        if (!validation.valid) {
            throw new Error(`Invalid workflow definition: ${validation.errors.map(e => e.message).join(', ')}`);
        }

        this.workflow = workflowDef;
        this.options = {
            ...options,
            // Support legacy option name
            gemini_api_key: options.gemini_api_key || options.geminiApiKey
        };
        this.tokens = [];
        this.contextManager = new ContextManager();
        this.status = 'idle';
        this.activityMap = new Map();
        this.edgeMap = new Map();
        this.decisionNodeMap = new Map();
        this.roleMap = new Map();
        this.decisionEvaluator = new DecisionEvaluator(workflowDef.decision_nodes);

        this.initializeMaps();
        this.initializeContexts();
    }

    private initializeMaps(): void {
        for (const activity of this.workflow.activities) {
            this.activityMap.set(activity.id, activity);
        }

        for (const edge of this.workflow.edges) {
            if (!this.edgeMap.has(edge.source_id)) {
                this.edgeMap.set(edge.source_id, []);
            }
            this.edgeMap.get(edge.source_id)?.push(edge);
        }

        for (const decisionNode of this.workflow.decision_nodes) {
            this.decisionNodeMap.set(decisionNode.id, decisionNode);
        }

        // Load roles from options and workflow
        if (this.options.roles) {
            for (const role of this.options.roles) {
                this.roleMap.set(role.id, role);
            }
        }
    }

    private initializeContexts(): void {
        for (const context of this.workflow.contexts) {
            this.contextManager.register(context);
        }
    }

    private log(message: string): void {
        if (this.options.verbose) {
            console.log(`[WorkflowEngine] ${message}`);
        }
    }

    public async start(initialData: Record<string, unknown> = {}): Promise<UUID> {
        if (this.status === 'running') {
            throw new Error('Workflow is already running');
        }

        if (this.workflow.activities.length === 0) {
            throw new Error('Workflow has no activities');
        }

        // Strategy: Find node with no incoming edges from Activities
        let startNodeId: UUID | undefined;

        const targetIds = new Set(this.workflow.edges.map(e => e.target_id));
        const startActivities = this.workflow.activities.filter(a => !targetIds.has(a.id));

        if (startActivities.length > 0) {
            startNodeId = startActivities[0].id;
        } else {
            // Cycle or just pick first
            startNodeId = this.workflow.activities[0].id;
        }

        // Add workflow metadata to initial data
        const tokenData = {
            ...initialData,
            _workflow_id: this.workflow.id,
            _workflow_name: this.workflow.name,
            _started_at: new Date().toISOString()
        };

        const token = new Token(startNodeId, tokenData, undefined, this.workflow.id);
        this.tokens.push(token);
        this.status = 'running';

        this.log(`Started workflow "${this.workflow.name}" with token ${token.id}`);

        return token.id;
    }

    public async runStep(): Promise<void> {
        if (this.status !== 'running' && this.status !== 'waiting_human') return;

        const activeTokens = this.tokens.filter(t => t.status === 'active');

        if (activeTokens.length === 0) {
            // Check if any tokens are waiting for human input
            const waitingTokens = this.tokens.filter(t => t.status === 'waiting');
            if (waitingTokens.length > 0) {
                this.status = 'waiting_human';
                return;
            }
            this.status = 'completed';
            return;
        }

        for (const token of activeTokens) {
            await this.processToken(token);
        }

        // Re-check status
        const allDone = this.tokens.every(t =>
            t.status === 'completed' || t.status === 'failed' || t.status === 'cancelled'
        );
        if (allDone) {
            this.status = 'completed';
        }
    }

    /**
     * Run workflow to completion (or until paused/failed)
     */
    public async run(maxSteps: number = 1000): Promise<EngineStatus> {
        let steps = 0;
        while (this.status === 'running' && steps < maxSteps) {
            await this.runStep();
            steps++;
        }
        this.log(`Workflow completed after ${steps} steps with status: ${this.status}`);
        return this.status;
    }

    private async processToken(token: Token): Promise<void> {
        const currentNodeId = token.activityId;

        // Check if this is a decision node
        const decisionNode = this.decisionNodeMap.get(currentNodeId);
        if (decisionNode) {
            await this.processDecisionNode(token, decisionNode);
            return;
        }

        // Process as activity
        const activity = this.activityMap.get(currentNodeId);
        if (!activity) {
            this.log(`Activity not found: ${currentNodeId}`);
            token.updateStatus('failed');
            return;
        }

        const startTime = new Date();

        try {
            // Create appropriate actor
            const actor = this.createActor(activity);

            if (actor) {
                // Add token metadata to inputs
                const inputs = {
                    ...token.contextData,
                    _token_id: token.id,
                    _workflow_id: this.workflow.id,
                    _activity_id: activity.id,
                    _activity_name: activity.name
                };

                this.log(`Executing activity "${activity.name}" with ${activity.actor_type} actor`);

                const output = await actor.execute(activity, inputs);

                const endTime = new Date();
                const processTime = calculate_duration(startTime, endTime);
                const analytics: Analytics = {
                    process_time: processTime,
                    cycle_time: processTime,
                    lead_time: processTime,
                    value_added: activity.analytics?.value_added ?? true,
                    waste_categories: []
                };

                // Check if human task requires waiting
                if (output._requires_human_action && this.options.wait_for_human_tasks) {
                    analytics.waste_categories?.push('waiting' as WasteCategory);
                    token.updateStatus('waiting', analytics);
                    token.mergeData({
                        ...output,
                        _waiting_since: endTime.toISOString()
                    });
                    this.log(`Token paused waiting for human task ${output._human_task_id}`);
                    return;
                }

                // Merge output back to token data
                token.mergeData(output);

                // Determine next node(s)
                await this.advanceToken(token, currentNodeId, analytics);

            } else {
                // No actor - just advance
                await this.advanceToken(token, currentNodeId);
            }

        } catch (error: unknown) {
            const endTime = new Date();
            const errorMessage = error instanceof Error ? error.message : String(error);
            const errorStack = error instanceof Error ? error.stack : undefined;
            console.error(`Error processing activity ${activity.name}:`, error);

            const analytics: Analytics = {
                process_time: calculate_duration(startTime, endTime),
                waste_categories: ['defects' as WasteCategory],
                error_rate: 1
            };

            token.updateStatus('failed', analytics);
            token.mergeData({ _error: errorMessage, _stack: errorStack });
        }
    }

    /**
     * Process a decision node
     */
    private async processDecisionNode(token: Token, decision: DecisionNode): Promise<void> {
        this.log(`Evaluating decision node "${decision.name}"`);

        const result = evaluate_decision(decision, token.contextData);

        this.log(`Decision result: matched=${result.matched}, outputs=${JSON.stringify(result.outputs)}`);

        // Merge decision outputs into token
        token.mergeData({
            _decision_node_id: decision.id,
            _decision_matched: result.matched,
            _decision_outputs: result.outputs
        });

        if (result.output_edge_id) {
            // Follow the specific edge from decision
            token.move(this.getEdgeTarget(result.output_edge_id) || result.output_edge_id);
        } else {
            // No matching edge - use default or fail
            if (decision.default_output_edge_id) {
                token.move(this.getEdgeTarget(decision.default_output_edge_id) || decision.default_output_edge_id);
            } else {
                this.log(`No matching rule and no default edge for decision "${decision.name}"`);
                token.updateStatus('failed');
                token.mergeData({ _error: 'No matching decision rule and no default path' });
            }
        }
    }

    /**
     * Get the target node ID for an edge ID
     */
    private getEdgeTarget(edgeId: UUID): UUID | undefined {
        for (const edge of this.workflow.edges) {
            if (edge.id === edgeId) {
                return edge.target_id;
            }
        }
        return undefined;
    }

    /**
     * Advance token to next node(s) based on outgoing edges
     */
    private async advanceToken(token: Token, currentNodeId: UUID, analytics?: Analytics): Promise<void> {
        const outgoingEdges = this.edgeMap.get(currentNodeId) || [];

        if (outgoingEdges.length === 0) {
            token.updateStatus('completed', analytics);
            this.log(`Token ${token.id} completed (no outgoing edges)`);
            return;
        }

        // Evaluate edge conditions to find the next edge
        const nextEdge = this.selectNextEdge(outgoingEdges, token.contextData);

        if (nextEdge) {
            token.move(nextEdge.target_id, analytics);
            this.log(`Token ${token.id} moved to ${nextEdge.target_id}`);
        } else {
            // No valid edge found
            token.updateStatus('failed', analytics);
            token.mergeData({ _error: 'No valid outgoing edge found' });
        }
    }

    /**
     * Select the next edge based on conditions
     */
    private selectNextEdge(edges: Edge[], context: Record<string, unknown>): Edge | undefined {
        // First, check for conditional edges
        for (const edge of edges) {
            if (edge.condition) {
                if (this.evaluateEdgeCondition(edge.condition, context)) {
                    return edge;
                }
            }
        }

        // Then, check for default edge
        const defaultEdge = edges.find(e => e.is_default);
        if (defaultEdge) {
            return defaultEdge;
        }

        // Finally, take first edge if no conditions match
        return edges[0];
    }

    /**
     * Evaluate a simple edge condition
     * Supports: property == value, property != value, property (truthy check)
     */
    private evaluateEdgeCondition(condition: string, context: Record<string, unknown>): boolean {
        try {
            // Handle equality check: property == value
            const eqMatch = condition.match(/^(\w+)\s*==\s*(.+)$/);
            if (eqMatch) {
                const [, prop, val] = eqMatch;
                const contextVal = context[prop];
                const compareVal = val.trim().replace(/^["']|["']$/g, '');
                return String(contextVal) === compareVal || contextVal === JSON.parse(val);
            }

            // Handle inequality check: property != value
            const neqMatch = condition.match(/^(\w+)\s*!=\s*(.+)$/);
            if (neqMatch) {
                const [, prop, val] = neqMatch;
                const contextVal = context[prop];
                const compareVal = val.trim().replace(/^["']|["']$/g, '');
                return String(contextVal) !== compareVal && contextVal !== JSON.parse(val);
            }

            // Handle greater than: property > value
            const gtMatch = condition.match(/^(\w+)\s*>\s*(\d+(?:\.\d+)?)$/);
            if (gtMatch) {
                const [, prop, val] = gtMatch;
                return Number(context[prop]) > Number(val);
            }

            // Handle less than: property < value
            const ltMatch = condition.match(/^(\w+)\s*<\s*(\d+(?:\.\d+)?)$/);
            if (ltMatch) {
                const [, prop, val] = ltMatch;
                return Number(context[prop]) < Number(val);
            }

            // Handle truthy check: just property name
            if (/^\w+$/.test(condition)) {
                return Boolean(context[condition]);
            }

            // Default: evaluate as-is (unsafe, but matches original TODO behavior)
            return Boolean(context[condition]);
        } catch {
            return false;
        }
    }

    /**
     * Create the appropriate actor for an activity
     */
    private createActor(activity: Activity): Actor | undefined {
        switch (activity.actor_type) {
            case 'application':
                return new SoftwareAgent();

            case 'ai_agent':
                if (!this.options.gemini_api_key) {
                    throw new Error('Gemini API Key required for AI Agent. Set gemini_api_key in options.');
                }
                const role = this.roleMap.get(activity.role_id);
                return new AIAgent(this.options.gemini_api_key, role);

            case 'robot':
                return new RobotAgent(this.options.robot_config);

            case 'human':
                return new HumanAgent({
                    wait_for_completion: false // Engine will handle waiting state
                });

            default:
                this.log(`Unknown actor type: ${activity.actor_type}`);
                return undefined;
        }
    }

    /**
     * Resume a token that was waiting for human input
     */
    public resumeToken(tokenId: UUID, output: Record<string, unknown>): boolean {
        const token = this.tokens.find(t => t.id === tokenId);
        if (!token || token.status !== 'waiting') {
            return false;
        }

        const waitingSince = token.getData('_waiting_since');
        let analytics: Analytics | undefined;

        if (waitingSince) {
            const waitTime = calculate_duration(new Date(waitingSince), new Date());
            analytics = {
                wait_time: waitTime,
                waste_categories: ['waiting' as WasteCategory]
            };
        }

        token.mergeData(output);
        token.updateStatus('active', analytics);

        if (this.status === 'waiting_human') {
            this.status = 'running';
        }

        return true;
    }

    public getStatus(): EngineStatus {
        return this.status;
    }

    public getContext(contextId: UUID): unknown {
        return this.contextManager.get(contextId);
    }

    public getTokens(): Token[] {
        return this.tokens;
    }

    public getWorkflow(): Workflow {
        return this.workflow;
    }
}
